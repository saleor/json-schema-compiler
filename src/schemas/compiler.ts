import * as AjvStandalone from "ajv/dist/standalone";
import type { AnySchema, AnySchemaObject } from "ajv";
import { compile } from "json-schema-to-typescript";
import { getAjv } from "./ajv";
import { type JSONValue } from "./types";

type JSONSchema4 = Parameters<typeof compile>[0];

const getAjvForCodegen = (definitionsJson?: AnySchema) => {
	const ajv = getAjv({
		coerceTypes: true,
		code: {
			source: true,
			esm: true,
			optimize: true,
		},
		loadSchema(uri) {
			// @todo is this safe?
			return fetch(uri).then((res) => res.json() as Promise<AnySchemaObject>);
		},
	});
	if (definitionsJson) {
		ajv.addSchema(definitionsJson);
	}
	return ajv;
};

export async function compileSchemaToTypes(
	schemaName: string,
	schema: JSONValue,
	definitionsDir: string,
) {
	const types = await compile(schema as JSONSchema4, schemaName, {
		unknownAny: true,
		cwd: definitionsDir,
	});
	const typesWithDefaultExport = [
		`import type { ValidateFunction } from "ajv";`,
		`import type { JSONObject } from '../../types';`,
		types,
		`declare const Validate${schemaName}: ValidateFunction<${schemaName}>;`,
		`export default Validate${schemaName};`,
	].join("\n");
	return typesWithDefaultExport;
}

export async function compileSchemaToJs(
	_schemaName: string,
	schema: JSONValue,
	definitionsJson?: AnySchema,
) {
	const ajv = getAjvForCodegen(definitionsJson);
	const validate = await ajv.compileAsync(schema as AnySchemaObject);
	const sourceCode = AjvStandalone.default(ajv, validate);
	return `/* c8 ignore start */\n` + sourceCode;
}
