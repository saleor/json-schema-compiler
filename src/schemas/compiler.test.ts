import { expect, describe, it } from "vitest";
import * as Prettier from "prettier";
import { compileSchemaToJs, compileSchemaToTypes } from "./compiler";

const exampleSchema = {
	$schema: "http://json-schema.org/draft-07/schema",
	type: "object",
	properties: {
		pspReference: { type: "string" },
		data: { type: "object", additionalProperties: true, properties: {} },
		result: {
			anyOf: [
				{ type: "string", const: "CHARGE_SUCCESS" },
				{ type: "string", const: "CHARGE_FAILURE" },
				{ type: "string", const: "CHARGE_REQUESTED" },
				{ type: "string", const: "AUTHORIZATION_SUCCESS" },
				{ type: "string", const: "AUTHORIZATION_FAILURE" },
				{ type: "string", const: "AUTHORIZATION_REQUESTED" },
			],
		},
		amount: { type: "number", minimum: 0 },
		time: { type: "string", format: "date-time" },
		externalUrl: { type: "string" },
		message: { type: "string" },
	},
	required: ["pspReference", "data", "result", "amount", "time", "externalUrl", "message"],
} as const;

describe("JSON schema compiler", () => {
	describe(`compileSchemaToJs`, () => {
		it(`compileSchemaToJs compiles schema to JS code`, async () => {
			const resultCode = await compileSchemaToJs("SchemaName", exampleSchema);
			const formattedCode = Prettier.format(resultCode, { parser: "typescript" });
			expect(formattedCode).toMatchSnapshot();
		});
	});

	describe(`compileSchemaToTypes`, () => {
		it(`compileSchemaToTypes compiles schema to TS types`, async () => {
			expect(await compileSchemaToTypes("SchemaName", exampleSchema, "")).toMatchInlineSnapshot(`
        "import type { ValidateFunction } from \\"ajv\\";
        import type { JSONObject } from '../../types';
        /* eslint-disable */
        /**
         * This file was automatically generated by json-schema-to-typescript.
         * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
         * and run json-schema-to-typescript to regenerate this file.
         */

        export interface SchemaName {
          pspReference: string;
          data: {
            [k: string]: unknown;
          };
          result:
            | \\"CHARGE_SUCCESS\\"
            | \\"CHARGE_FAILURE\\"
            | \\"CHARGE_REQUESTED\\"
            | \\"AUTHORIZATION_SUCCESS\\"
            | \\"AUTHORIZATION_FAILURE\\"
            | \\"AUTHORIZATION_REQUESTED\\";
          amount: number;
          time: string;
          externalUrl: string;
          message: string;
          [k: string]: unknown;
        }

        declare const ValidateSchemaName: ValidateFunction<SchemaName>;
        export default ValidateSchemaName;"
      `);
		});
	});
});
