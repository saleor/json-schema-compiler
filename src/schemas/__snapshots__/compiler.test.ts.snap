// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JSON schema compiler > compileSchemaToJs > compileSchemaToJs compiles schema to JS code 1`] = `
"/* c8 ignore start */
\\"use strict\\";
export const validate = validate10;
export default validate10;
const schema11 = {
  $schema: \\"http://json-schema.org/draft-07/schema\\",
  type: \\"object\\",
  properties: {
    pspReference: { type: \\"string\\" },
    data: { type: \\"object\\", additionalProperties: true, properties: {} },
    result: {
      anyOf: [
        { type: \\"string\\", const: \\"CHARGE_SUCCESS\\" },
        { type: \\"string\\", const: \\"CHARGE_FAILURE\\" },
        { type: \\"string\\", const: \\"CHARGE_REQUESTED\\" },
        { type: \\"string\\", const: \\"AUTHORIZATION_SUCCESS\\" },
        { type: \\"string\\", const: \\"AUTHORIZATION_FAILURE\\" },
        { type: \\"string\\", const: \\"AUTHORIZATION_REQUESTED\\" },
      ],
    },
    amount: { type: \\"number\\", minimum: 0 },
    time: { type: \\"string\\", format: \\"date-time\\" },
    externalUrl: { type: \\"string\\" },
    message: { type: \\"string\\" },
  },
  required: [
    \\"pspReference\\",
    \\"data\\",
    \\"result\\",
    \\"amount\\",
    \\"time\\",
    \\"externalUrl\\",
    \\"message\\",
  ],
};
const formats0 = require(\\"ajv-formats/dist/formats\\").fullFormats[\\"date-time\\"];
function validate10(
  data,
  { instancePath = \\"\\", parentData, parentDataProperty, rootData = data } = {}
) {
  let vErrors = null;
  let errors = 0;
  if (errors === 0) {
    if (data && typeof data == \\"object\\" && !Array.isArray(data)) {
      let missing0;
      if (
        (data.pspReference === undefined && (missing0 = \\"pspReference\\")) ||
        (data.data === undefined && (missing0 = \\"data\\")) ||
        (data.result === undefined && (missing0 = \\"result\\")) ||
        (data.amount === undefined && (missing0 = \\"amount\\")) ||
        (data.time === undefined && (missing0 = \\"time\\")) ||
        (data.externalUrl === undefined && (missing0 = \\"externalUrl\\")) ||
        (data.message === undefined && (missing0 = \\"message\\"))
      ) {
        validate10.errors = [
          {
            instancePath,
            schemaPath: \\"#/required\\",
            keyword: \\"required\\",
            params: { missingProperty: missing0 },
            message: \\"must have required property '\\" + missing0 + \\"'\\",
          },
        ];
        return false;
      } else {
        if (data.pspReference !== undefined) {
          let data0 = data.pspReference;
          const _errs1 = errors;
          if (typeof data0 !== \\"string\\") {
            let dataType0 = typeof data0;
            let coerced0 = undefined;
            if (!(coerced0 !== undefined)) {
              if (dataType0 == \\"number\\" || dataType0 == \\"boolean\\") {
                coerced0 = \\"\\" + data0;
              } else if (data0 === null) {
                coerced0 = \\"\\";
              } else {
                validate10.errors = [
                  {
                    instancePath: instancePath + \\"/pspReference\\",
                    schemaPath: \\"#/properties/pspReference/type\\",
                    keyword: \\"type\\",
                    params: { type: \\"string\\" },
                    message: \\"must be string\\",
                  },
                ];
                return false;
              }
            }
            if (coerced0 !== undefined) {
              data0 = coerced0;
              if (data !== undefined) {
                data[\\"pspReference\\"] = coerced0;
              }
            }
          }
          var valid0 = _errs1 === errors;
        } else {
          var valid0 = true;
        }
        if (valid0) {
          if (data.data !== undefined) {
            let data1 = data.data;
            const _errs3 = errors;
            if (errors === _errs3) {
              if (data1 && typeof data1 == \\"object\\" && !Array.isArray(data1)) {
              } else {
                validate10.errors = [
                  {
                    instancePath: instancePath + \\"/data\\",
                    schemaPath: \\"#/properties/data/type\\",
                    keyword: \\"type\\",
                    params: { type: \\"object\\" },
                    message: \\"must be object\\",
                  },
                ];
                return false;
              }
            }
            var valid0 = _errs3 === errors;
          } else {
            var valid0 = true;
          }
          if (valid0) {
            if (data.result !== undefined) {
              let data2 = data.result;
              const _errs6 = errors;
              const _errs7 = errors;
              let valid1 = false;
              const _errs8 = errors;
              if (typeof data2 !== \\"string\\") {
                let dataType1 = typeof data2;
                let coerced1 = undefined;
                if (!(coerced1 !== undefined)) {
                  if (dataType1 == \\"number\\" || dataType1 == \\"boolean\\") {
                    coerced1 = \\"\\" + data2;
                  } else if (data2 === null) {
                    coerced1 = \\"\\";
                  } else {
                    const err0 = {
                      instancePath: instancePath + \\"/result\\",
                      schemaPath: \\"#/properties/result/anyOf/0/type\\",
                      keyword: \\"type\\",
                      params: { type: \\"string\\" },
                      message: \\"must be string\\",
                    };
                    if (vErrors === null) {
                      vErrors = [err0];
                    } else {
                      vErrors.push(err0);
                    }
                    errors++;
                  }
                }
                if (coerced1 !== undefined) {
                  data2 = coerced1;
                  if (data !== undefined) {
                    data[\\"result\\"] = coerced1;
                  }
                }
              }
              if (\\"CHARGE_SUCCESS\\" !== data2) {
                const err1 = {
                  instancePath: instancePath + \\"/result\\",
                  schemaPath: \\"#/properties/result/anyOf/0/const\\",
                  keyword: \\"const\\",
                  params: { allowedValue: \\"CHARGE_SUCCESS\\" },
                  message: \\"must be equal to constant\\",
                };
                if (vErrors === null) {
                  vErrors = [err1];
                } else {
                  vErrors.push(err1);
                }
                errors++;
              }
              var _valid0 = _errs8 === errors;
              valid1 = valid1 || _valid0;
              if (!valid1) {
                const _errs10 = errors;
                if (typeof data2 !== \\"string\\") {
                  let dataType2 = typeof data2;
                  let coerced2 = undefined;
                  if (!(coerced2 !== undefined)) {
                    if (dataType2 == \\"number\\" || dataType2 == \\"boolean\\") {
                      coerced2 = \\"\\" + data2;
                    } else if (data2 === null) {
                      coerced2 = \\"\\";
                    } else {
                      const err2 = {
                        instancePath: instancePath + \\"/result\\",
                        schemaPath: \\"#/properties/result/anyOf/1/type\\",
                        keyword: \\"type\\",
                        params: { type: \\"string\\" },
                        message: \\"must be string\\",
                      };
                      if (vErrors === null) {
                        vErrors = [err2];
                      } else {
                        vErrors.push(err2);
                      }
                      errors++;
                    }
                  }
                  if (coerced2 !== undefined) {
                    data2 = coerced2;
                    if (data !== undefined) {
                      data[\\"result\\"] = coerced2;
                    }
                  }
                }
                if (\\"CHARGE_FAILURE\\" !== data2) {
                  const err3 = {
                    instancePath: instancePath + \\"/result\\",
                    schemaPath: \\"#/properties/result/anyOf/1/const\\",
                    keyword: \\"const\\",
                    params: { allowedValue: \\"CHARGE_FAILURE\\" },
                    message: \\"must be equal to constant\\",
                  };
                  if (vErrors === null) {
                    vErrors = [err3];
                  } else {
                    vErrors.push(err3);
                  }
                  errors++;
                }
                var _valid0 = _errs10 === errors;
                valid1 = valid1 || _valid0;
                if (!valid1) {
                  const _errs12 = errors;
                  if (typeof data2 !== \\"string\\") {
                    let dataType3 = typeof data2;
                    let coerced3 = undefined;
                    if (!(coerced3 !== undefined)) {
                      if (dataType3 == \\"number\\" || dataType3 == \\"boolean\\") {
                        coerced3 = \\"\\" + data2;
                      } else if (data2 === null) {
                        coerced3 = \\"\\";
                      } else {
                        const err4 = {
                          instancePath: instancePath + \\"/result\\",
                          schemaPath: \\"#/properties/result/anyOf/2/type\\",
                          keyword: \\"type\\",
                          params: { type: \\"string\\" },
                          message: \\"must be string\\",
                        };
                        if (vErrors === null) {
                          vErrors = [err4];
                        } else {
                          vErrors.push(err4);
                        }
                        errors++;
                      }
                    }
                    if (coerced3 !== undefined) {
                      data2 = coerced3;
                      if (data !== undefined) {
                        data[\\"result\\"] = coerced3;
                      }
                    }
                  }
                  if (\\"CHARGE_REQUESTED\\" !== data2) {
                    const err5 = {
                      instancePath: instancePath + \\"/result\\",
                      schemaPath: \\"#/properties/result/anyOf/2/const\\",
                      keyword: \\"const\\",
                      params: { allowedValue: \\"CHARGE_REQUESTED\\" },
                      message: \\"must be equal to constant\\",
                    };
                    if (vErrors === null) {
                      vErrors = [err5];
                    } else {
                      vErrors.push(err5);
                    }
                    errors++;
                  }
                  var _valid0 = _errs12 === errors;
                  valid1 = valid1 || _valid0;
                  if (!valid1) {
                    const _errs14 = errors;
                    if (typeof data2 !== \\"string\\") {
                      let dataType4 = typeof data2;
                      let coerced4 = undefined;
                      if (!(coerced4 !== undefined)) {
                        if (dataType4 == \\"number\\" || dataType4 == \\"boolean\\") {
                          coerced4 = \\"\\" + data2;
                        } else if (data2 === null) {
                          coerced4 = \\"\\";
                        } else {
                          const err6 = {
                            instancePath: instancePath + \\"/result\\",
                            schemaPath: \\"#/properties/result/anyOf/3/type\\",
                            keyword: \\"type\\",
                            params: { type: \\"string\\" },
                            message: \\"must be string\\",
                          };
                          if (vErrors === null) {
                            vErrors = [err6];
                          } else {
                            vErrors.push(err6);
                          }
                          errors++;
                        }
                      }
                      if (coerced4 !== undefined) {
                        data2 = coerced4;
                        if (data !== undefined) {
                          data[\\"result\\"] = coerced4;
                        }
                      }
                    }
                    if (\\"AUTHORIZATION_SUCCESS\\" !== data2) {
                      const err7 = {
                        instancePath: instancePath + \\"/result\\",
                        schemaPath: \\"#/properties/result/anyOf/3/const\\",
                        keyword: \\"const\\",
                        params: { allowedValue: \\"AUTHORIZATION_SUCCESS\\" },
                        message: \\"must be equal to constant\\",
                      };
                      if (vErrors === null) {
                        vErrors = [err7];
                      } else {
                        vErrors.push(err7);
                      }
                      errors++;
                    }
                    var _valid0 = _errs14 === errors;
                    valid1 = valid1 || _valid0;
                    if (!valid1) {
                      const _errs16 = errors;
                      if (typeof data2 !== \\"string\\") {
                        let dataType5 = typeof data2;
                        let coerced5 = undefined;
                        if (!(coerced5 !== undefined)) {
                          if (dataType5 == \\"number\\" || dataType5 == \\"boolean\\") {
                            coerced5 = \\"\\" + data2;
                          } else if (data2 === null) {
                            coerced5 = \\"\\";
                          } else {
                            const err8 = {
                              instancePath: instancePath + \\"/result\\",
                              schemaPath: \\"#/properties/result/anyOf/4/type\\",
                              keyword: \\"type\\",
                              params: { type: \\"string\\" },
                              message: \\"must be string\\",
                            };
                            if (vErrors === null) {
                              vErrors = [err8];
                            } else {
                              vErrors.push(err8);
                            }
                            errors++;
                          }
                        }
                        if (coerced5 !== undefined) {
                          data2 = coerced5;
                          if (data !== undefined) {
                            data[\\"result\\"] = coerced5;
                          }
                        }
                      }
                      if (\\"AUTHORIZATION_FAILURE\\" !== data2) {
                        const err9 = {
                          instancePath: instancePath + \\"/result\\",
                          schemaPath: \\"#/properties/result/anyOf/4/const\\",
                          keyword: \\"const\\",
                          params: { allowedValue: \\"AUTHORIZATION_FAILURE\\" },
                          message: \\"must be equal to constant\\",
                        };
                        if (vErrors === null) {
                          vErrors = [err9];
                        } else {
                          vErrors.push(err9);
                        }
                        errors++;
                      }
                      var _valid0 = _errs16 === errors;
                      valid1 = valid1 || _valid0;
                      if (!valid1) {
                        const _errs18 = errors;
                        if (typeof data2 !== \\"string\\") {
                          let dataType6 = typeof data2;
                          let coerced6 = undefined;
                          if (!(coerced6 !== undefined)) {
                            if (
                              dataType6 == \\"number\\" ||
                              dataType6 == \\"boolean\\"
                            ) {
                              coerced6 = \\"\\" + data2;
                            } else if (data2 === null) {
                              coerced6 = \\"\\";
                            } else {
                              const err10 = {
                                instancePath: instancePath + \\"/result\\",
                                schemaPath: \\"#/properties/result/anyOf/5/type\\",
                                keyword: \\"type\\",
                                params: { type: \\"string\\" },
                                message: \\"must be string\\",
                              };
                              if (vErrors === null) {
                                vErrors = [err10];
                              } else {
                                vErrors.push(err10);
                              }
                              errors++;
                            }
                          }
                          if (coerced6 !== undefined) {
                            data2 = coerced6;
                            if (data !== undefined) {
                              data[\\"result\\"] = coerced6;
                            }
                          }
                        }
                        if (\\"AUTHORIZATION_REQUESTED\\" !== data2) {
                          const err11 = {
                            instancePath: instancePath + \\"/result\\",
                            schemaPath: \\"#/properties/result/anyOf/5/const\\",
                            keyword: \\"const\\",
                            params: { allowedValue: \\"AUTHORIZATION_REQUESTED\\" },
                            message: \\"must be equal to constant\\",
                          };
                          if (vErrors === null) {
                            vErrors = [err11];
                          } else {
                            vErrors.push(err11);
                          }
                          errors++;
                        }
                        var _valid0 = _errs18 === errors;
                        valid1 = valid1 || _valid0;
                      }
                    }
                  }
                }
              }
              if (!valid1) {
                const err12 = {
                  instancePath: instancePath + \\"/result\\",
                  schemaPath: \\"#/properties/result/anyOf\\",
                  keyword: \\"anyOf\\",
                  params: {},
                  message: \\"must match a schema in anyOf\\",
                };
                if (vErrors === null) {
                  vErrors = [err12];
                } else {
                  vErrors.push(err12);
                }
                errors++;
                validate10.errors = vErrors;
                return false;
              } else {
                errors = _errs7;
                if (vErrors !== null) {
                  if (_errs7) {
                    vErrors.length = _errs7;
                  } else {
                    vErrors = null;
                  }
                }
              }
              var valid0 = _errs6 === errors;
            } else {
              var valid0 = true;
            }
            if (valid0) {
              if (data.amount !== undefined) {
                let data3 = data.amount;
                const _errs20 = errors;
                if (!(typeof data3 == \\"number\\" && isFinite(data3))) {
                  let dataType7 = typeof data3;
                  let coerced7 = undefined;
                  if (!(coerced7 !== undefined)) {
                    if (
                      dataType7 == \\"boolean\\" ||
                      data3 === null ||
                      (dataType7 == \\"string\\" && data3 && data3 == +data3)
                    ) {
                      coerced7 = +data3;
                    } else {
                      validate10.errors = [
                        {
                          instancePath: instancePath + \\"/amount\\",
                          schemaPath: \\"#/properties/amount/type\\",
                          keyword: \\"type\\",
                          params: { type: \\"number\\" },
                          message: \\"must be number\\",
                        },
                      ];
                      return false;
                    }
                  }
                  if (coerced7 !== undefined) {
                    data3 = coerced7;
                    if (data !== undefined) {
                      data[\\"amount\\"] = coerced7;
                    }
                  }
                }
                if (errors === _errs20) {
                  if (typeof data3 == \\"number\\" && isFinite(data3)) {
                    if (data3 < 0 || isNaN(data3)) {
                      validate10.errors = [
                        {
                          instancePath: instancePath + \\"/amount\\",
                          schemaPath: \\"#/properties/amount/minimum\\",
                          keyword: \\"minimum\\",
                          params: { comparison: \\">=\\", limit: 0 },
                          message: \\"must be >= 0\\",
                        },
                      ];
                      return false;
                    }
                  }
                }
                var valid0 = _errs20 === errors;
              } else {
                var valid0 = true;
              }
              if (valid0) {
                if (data.time !== undefined) {
                  let data4 = data.time;
                  const _errs22 = errors;
                  if (typeof data4 !== \\"string\\") {
                    let dataType8 = typeof data4;
                    let coerced8 = undefined;
                    if (!(coerced8 !== undefined)) {
                      if (dataType8 == \\"number\\" || dataType8 == \\"boolean\\") {
                        coerced8 = \\"\\" + data4;
                      } else if (data4 === null) {
                        coerced8 = \\"\\";
                      } else {
                        validate10.errors = [
                          {
                            instancePath: instancePath + \\"/time\\",
                            schemaPath: \\"#/properties/time/type\\",
                            keyword: \\"type\\",
                            params: { type: \\"string\\" },
                            message: \\"must be string\\",
                          },
                        ];
                        return false;
                      }
                    }
                    if (coerced8 !== undefined) {
                      data4 = coerced8;
                      if (data !== undefined) {
                        data[\\"time\\"] = coerced8;
                      }
                    }
                  }
                  if (errors === _errs22) {
                    if (errors === _errs22) {
                      if (typeof data4 === \\"string\\") {
                        if (!formats0.validate(data4)) {
                          validate10.errors = [
                            {
                              instancePath: instancePath + \\"/time\\",
                              schemaPath: \\"#/properties/time/format\\",
                              keyword: \\"format\\",
                              params: { format: \\"date-time\\" },
                              message:
                                'must match format \\"' + \\"date-time\\" + '\\"',
                            },
                          ];
                          return false;
                        }
                      }
                    }
                  }
                  var valid0 = _errs22 === errors;
                } else {
                  var valid0 = true;
                }
                if (valid0) {
                  if (data.externalUrl !== undefined) {
                    let data5 = data.externalUrl;
                    const _errs24 = errors;
                    if (typeof data5 !== \\"string\\") {
                      let dataType9 = typeof data5;
                      let coerced9 = undefined;
                      if (!(coerced9 !== undefined)) {
                        if (dataType9 == \\"number\\" || dataType9 == \\"boolean\\") {
                          coerced9 = \\"\\" + data5;
                        } else if (data5 === null) {
                          coerced9 = \\"\\";
                        } else {
                          validate10.errors = [
                            {
                              instancePath: instancePath + \\"/externalUrl\\",
                              schemaPath: \\"#/properties/externalUrl/type\\",
                              keyword: \\"type\\",
                              params: { type: \\"string\\" },
                              message: \\"must be string\\",
                            },
                          ];
                          return false;
                        }
                      }
                      if (coerced9 !== undefined) {
                        data5 = coerced9;
                        if (data !== undefined) {
                          data[\\"externalUrl\\"] = coerced9;
                        }
                      }
                    }
                    var valid0 = _errs24 === errors;
                  } else {
                    var valid0 = true;
                  }
                  if (valid0) {
                    if (data.message !== undefined) {
                      let data6 = data.message;
                      const _errs26 = errors;
                      if (typeof data6 !== \\"string\\") {
                        let dataType10 = typeof data6;
                        let coerced10 = undefined;
                        if (!(coerced10 !== undefined)) {
                          if (
                            dataType10 == \\"number\\" ||
                            dataType10 == \\"boolean\\"
                          ) {
                            coerced10 = \\"\\" + data6;
                          } else if (data6 === null) {
                            coerced10 = \\"\\";
                          } else {
                            validate10.errors = [
                              {
                                instancePath: instancePath + \\"/message\\",
                                schemaPath: \\"#/properties/message/type\\",
                                keyword: \\"type\\",
                                params: { type: \\"string\\" },
                                message: \\"must be string\\",
                              },
                            ];
                            return false;
                          }
                        }
                        if (coerced10 !== undefined) {
                          data6 = coerced10;
                          if (data !== undefined) {
                            data[\\"message\\"] = coerced10;
                          }
                        }
                      }
                      var valid0 = _errs26 === errors;
                    } else {
                      var valid0 = true;
                    }
                  }
                }
              }
            }
          }
        }
      }
    } else {
      validate10.errors = [
        {
          instancePath,
          schemaPath: \\"#/type\\",
          keyword: \\"type\\",
          params: { type: \\"object\\" },
          message: \\"must be object\\",
        },
      ];
      return false;
    }
  }
  validate10.errors = vErrors;
  return errors === 0;
}
"
`;
